feat(lgciu/sd): add LGCIU bus output, modify WHEEL page for LGCIUs

lukecologne opened the pull request:
<!-- ⚠⚠ Do not delete this pull request template! ⚠⚠ -->
<!-- Pull requests that do not follow this template are likely to be ignored. -->

## Summary of Changes
<!-- Please provide a summary of changes for this pull request, ensuring all changes are explained. -->
This PR:
* Adds (part of) the LGCIU output bus, for a list of bits in the discrete words and their description, see the Lvar docs.
   * Discrete words 2 and 4 and some bits in discrete word 3 have not been added, since the functionality for them is not currently implemented in the LGCIU.
* Modifies the WHEEL SD page to use the LGCIU output bus, and improve the indications.
   * LGCIU failures should now be properly handled, and the "door partially opened" case has been added to the door indications. 
  
   

## Screenshots (if necessary)
<!-- If your PR includes visual changes, screenshots from before and after your change should always be included. -->
<!-- Please make your best efforts to provide useful before and after screenshots. They should match camera angle, zoom, size, time of day, etc. -->

## References
<!-- You should be making changes based on some kind of a reference (manuals, videos, IRL photos). P3D/xplane/fsx references will only be accepted if we believe that one cannot reasonably obtain a better source. Please post screenshots of the references you used. Ask around in the discord for how to find references for what you are working on. Exceptions will probably be made for IRL A320 pilots and engineers. -->

<!-- If you are making a pull request related to the MCDU, please make sure you are ONLY referencing the Honeywell Pegasus Step 1A (Rev 0), 2009 edition manual. -->
<!-- If you do not have this manual, please ask on our discord for assistance -->

## Additional context
<!-- Add any other context about the pull request here. -->

<!-- You may optionally provide your discord username, so that we may contact you directly about the issue. -->
Discord username (if different from GitHub):

## Testing instructions
<!-- Detail how this PR should be tested by QA. Try to list important items that need checking, either directly changed by this PR or that could be affected by it -->
Check that the WHEEL page works correctly in all situations, including failures

Testing suggestion from Davy:
```
Use gravity gear extension on ground and in flight. If clicked and hold it should:

release doors on ground
release doors and gears in flight.
After use even with green system available, gear should be inoperative.

If clicked and held again, and if green system available it should:

close doors on ground
close doors and gears in flight. (Might need to recycle gear lever if LGCIU error)
```

<!-- DO NOT DELETE THIS -->
## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page

tracernz commented on src/instruments/src/SD/Pages/Wheel/Wheel.tsx:
```diff
@@ -178,114 +232,241 @@ type GearLocation = 'left' | 'center' | 'right';
+        lgciu2GearNotUplocked = lgciu2DiscreteWord3.getBitValue(11);
+        lgciu1GearDownlocked = lgciu1DiscreteWord1.getBitValue(23);
+        lgciu2GearDownlocked = lgciu2DiscreteWord1.getBitValue(23);
+        upLockFlagShown = upLockFlagShown && lgciu1DiscreteWord1.getBitValue(20) && lgciu1DiscreteWord1.getBitValue(20);
```
Should be lgciu2 for the second one?

lukecologne replied:
Oh yeah, nice copy paste error. Will fix shortly

tracernz commented on src/instruments/src/SD/Pages/Wheel/Wheel.tsx:
```diff
@@ -178,114 +232,241 @@ type GearLocation = 'left' | 'center' | 'right';
+        lgciu2GearNotUplocked = lgciu2DiscreteWord3.getBitValue(12);
+        lgciu1GearDownlocked = lgciu1DiscreteWord1.getBitValue(24);
+        lgciu2GearDownlocked = lgciu2DiscreteWord1.getBitValue(24);
+        upLockFlagShown = upLockFlagShown && lgciu1DiscreteWord1.getBitValue(21) && lgciu1DiscreteWord1.getBitValue(21);
```
Should be lgciu2 for the second one?

tracernz commented on src/instruments/src/SD/Pages/Wheel/Wheel.tsx:
```diff
@@ -178,114 +232,241 @@ type GearLocation = 'left' | 'center' | 'right';
+        lgciu2GearNotUplocked = lgciu2DiscreteWord3.getBitValue(13);
+        lgciu1GearDownlocked = lgciu1DiscreteWord1.getBitValue(25);
+        lgciu2GearDownlocked = lgciu2DiscreteWord1.getBitValue(25);
+        upLockFlagShown = upLockFlagShown && lgciu1DiscreteWord1.getBitValue(22) && lgciu1DiscreteWord1.getBitValue(22);
```
Should be lgciu2 for the second one?

crocket63 commented on src/systems/systems/src/landing_gear/mod.rs:
```diff
@@ -744,6 +756,131 @@ impl LandingGearControlInterfaceUnit {
+            set_arinc429_bit(
+                &mut word,
+                11,
+                !self.sensor_inputs.left_gear_up_and_locked && !self.gear_handle_is_down(),
```
I'm wondering if we should use sensor_inputs.left_gear_up_and_locked directly or use access methods that are also handling the value if failed/unpowered. Right now probably won't change anything as you just send invalid frame when unpowered, but something to remember to check when adding more things I believe.

crocket63 reviewed (changes requested):
Lgtm, however couldn't we add one or two tests to check per bit write/read in arinc429.rs ?

lukecologne commented on src/systems/systems/src/landing_gear/mod.rs:
```diff
@@ -744,6 +756,131 @@ impl LandingGearControlInterfaceUnit {
+
+    fn discrete_word_1(&self) -> Arinc429Word<f32> {
+        if !self.is_powered {
+            Arinc429Word::new(0., SignStatus::FailureWarning)
```
Yes, IRL a power failure would stop the transmission of the source computer, and the receiving computers would then register that and process that information accordingly. Some set the received label internally to Failure Warning, others do it differently. For now, we just set the SSM to FW at the source, that should be an alright approximation.

lukecologne commented:
I've added a bit-get method, and also added a test for both (similar to the existing one, testing the get and set at the same time. Hopefully it's adequate

crocket63 reviewed (approved):
Go go go

davidwalschots commented on src/systems/systems/src/landing_gear/mod.rs:
```diff
@@ -797,6 +898,10 @@ impl SimulationElement for LandingGearControlInterfaceUnit {
 
         writer.write(&self.fault_ecam_id, self.status() != LgciuStatus::Ok);
+
+        writer.write(&self.discrete_word_1_id, f64::from(self.discrete_word_1()));
+        writer.write(&self.discrete_word_2_id, f64::from(self.discrete_word_2()));
+        writer.write(&self.discrete_word_3_id, f64::from(self.discrete_word_3()));
```
**suggestion**: Add a `Writer` implementation for `Arinc429Word<f32>` which then calls the appropriate `From` implementation:

```rust
impl<T: Writer> Write<Arinc429Word<f32>> for T {
    fn convert(&mut self, value: Arinc429Word<f32>) -> f64 {
        value.into()
    }
}
```

```suggestion
        writer.write(&self.discrete_word_1_id, self.discrete_word_1());
        writer.write(&self.discrete_word_2_id, self.discrete_word_2());
        writer.write(&self.discrete_word_3_id, self.discrete_word_3());
```

davidwalschots commented on src/systems/systems/src/landing_gear/mod.rs:
```diff
@@ -744,6 +756,95 @@ impl LandingGearControlInterfaceUnit {
+            word.set_bit(18, !self.sensor_inputs.right_door_up_and_locked);
+            word.set_bit(19, !self.sensor_inputs.nose_door_up_and_locked);
+            // The conditions for bits 20, 21 and 22 are not implemented, so they are set to false for now.
+            word.set_bit(20, false);
+            word.set_bit(21, false);
+            word.set_bit(22, false);
+            word.set_bit(23, self.sensor_inputs.downlock_state(GearWheel::LEFT));
+            word.set_bit(24, self.sensor_inputs.downlock_state(GearWheel::RIGHT));
+            word.set_bit(25, self.sensor_inputs.downlock_state(GearWheel::NOSE));
+            word.set_bit(26, self.sensor_inputs.left_gear_compressed(true));
+            word.set_bit(27, self.sensor_inputs.right_gear_compressed(true));
+            word.set_bit(28, self.sensor_inputs.nose_gear_compressed(true));
+            word.set_bit(29, self.gear_handle_is_down());
```
**suggestion**: Consider creating a declarative macro to reduce the amount of ` word.set_bit` repetition. Not sure if there's some funky thing we can do here, might be worth looking into.

lukecologne replied:
Not sure how to do that, any suggestions?

davidwalschots commented on src/systems/systems/src/shared/arinc429.rs:
```diff
@@ -40,6 +40,35 @@ impl<T: Copy> Arinc429Word<T> {
         self.ssm == SignStatus::NormalOperation
     }
 }
+impl Arinc429Word<f32> {
+    pub fn set_bit(&mut self, bit: u32, value: bool) {
+        self.value =
+            (((self.value as u32) & !(1 << (bit - 1))) | ((value as u32) << (bit - 1))) as f32;
+    }
+
+    pub fn get_bit(&self, bit: u32) -> bool {
+        ((self.value as u32 >> (bit - 1)) & 1) != 0
+    }
+}
```
**thought**: I think we can store the value as `u32` instead of `f32`? I see a lot of conversion between these two numeric types, but am not really seeing a reason for doing so.

davidwalschots commented on src/systems/systems/src/shared/arinc429.rs:
```diff
@@ -40,6 +40,35 @@ impl<T: Copy> Arinc429Word<T> {
     }
 }
+impl Arinc429Word<f32> {
+    pub fn set_bit(&mut self, bit: u32, value: bool) {
```
**suggestion**: I think we will not exceed 256 bits within set_bit. Therefore this can be a `u8`.
```suggestion
    pub fn set_bit(&mut self, bit: u8, value: bool) {
```

**suggestion**: Add `debug_assert!` to limit the valid usage to between 1 and 32 (or 0 and 31, depending on the standard).

lukecologne replied:
Implemented the first suggestion, but why only `debug_assert`?

lukecologne replied:
Alright, added the `debug_assert` for the bit range that is possible with the ARINC 429 data word structure (bits 11 to 29).

davidwalschots commented on src/systems/systems/src/shared/arinc429.rs:
```diff
@@ -40,6 +40,35 @@ impl<T: Copy> Arinc429Word<T> {
+        ((self.value as u32 >> (bit - 1)) & 1) != 0
+    }
+}
+impl From<f64> for Arinc429Word<f32> {
```
**thought**: This implementation was probably created in anticipation of future reading of the value from a simvar.  **suggestion**: Add a `Reader` implementation as well (next to the `Writer`).

davidwalschots commented on src/systems/systems/src/landing_gear/mod.rs:
```diff
@@ -744,6 +756,95 @@ impl LandingGearControlInterfaceUnit {
+
+    fn discrete_word_2(&self) -> Arinc429Word<f32> {
+        if !self.is_powered {
+            Arinc429Word::new(0., SignStatus::FailureWarning)
+        } else {
+            let mut word = Arinc429Word::new(0., SignStatus::NormalOperation);
```
**suggestion**: Consider creating another constructor function which doesn't take a value, as it seems these usages always set the value to `0.`.

```rust
impl Arinc429Word<u32> {
    pub fn bits(ssm: SignStatus) -> Self {
        Self { value: 0., ssm }
    }
}
```

lukecologne replied:
I think it's better to keep it more explicit, so you can directly see it initializes to 0 and don't have to look up what the constructor does. Also, IMO, it wouldn't really save much to write anyways

2hwk commented:
Testing instructions could be a bit more detailed, it's hard to define what "behaving correctly" means

crocket63 commented:
> Testing instructions could be a bit more detailed, it's hard to define what "behaving correctly" means, a list of test cases and expected outputs would be helpful

Can use emergency extension for a nice test case:

Use gravity gear extension on ground and in flight. If clicked and hold it should:

release doors on ground
release doors and gears in flight.
After use even with green system available, gear should be inoperative.

If clicked and held again, and if green system available it should:

close doors on ground
close doors and gears in flight. (Might need to recycle gear lever if LGCIU error)

beheh commented on src/systems/systems/src/landing_gear/mod.rs:
```diff
@@ -744,6 +756,95 @@ impl LandingGearControlInterfaceUnit {
         self.status
     }
+
+    fn discrete_word_1(&self) -> Arinc429Word<u32> {
```
Would you mind making these accessible to other systems (like the FWC in future?). You could either make the method public, or add it to a public trait and implement that (e.g. like `systems::shared::LgciuWeightOnWheels`);

lukecologne replied:
I'll just make it public, I don't see a use of making it a trait, since it's not going to be different on different aircraft, or am I missing something?
